### PR TITLE
RoPE with frequencies as optional input

### PR DESCRIPTION
**Pull request by awni**
This enables alternative ropes like Yarn, RoPE with fine-tuned parameters, NTK (used in llama 3.1), Su RopE (long-context phi3) to use the `mx.fast.rope` op.

### CODE REVIEW

**Comment by awni** [comment]
Relevant PR in MLX examples https://github.com/ml-explore/mlx-examples/pull/945

**Comment by awni** [comment]
### Benchmarks

Timings on M2 Ultra, all models are 4-bit.

Model       | Slow RoPE | Fast RoPE
----- | ----- | ------
Llama 3.1 8B | 86.6 | 105.2
Phi-3 128k | 102.9 | 173.0
DeepSeekV2 Chat Lite | 100.3 | 111.5

File: mlx/backend/metal/rope.cpp [hunk 1]
```diff
@@ -1,4 +1,5 @@
 // Copyright © 2023-2024 Apple Inc.
+#include <iostream> // TODO
```
Comment (angeloskath) [inline mlx/backend/metal/rope.cpp:2]: Forgotten.

**Review by angeloskath — APPROVED** [review]
Looks great! 🚀

File: python/src/fast.cpp [hunk 2]
```diff
@@ -82,23 +82,25 @@ void init_fast(nb::module_& parent_module) {
...
 
         Args:
             a (array): Input array.
             dims (int): The feature dimensions to be rotated. If the input feature
-                is larger than dims then the rest is left unchanged.
+              is larger than dims then the rest is left unchanged.
             traditional (bool): If set to ``True`` choose the traditional
-                implementation which rotates consecutive dimensions.
+              implementation which rotates consecutive dimensions.
             base (float): The base used to compute angular frequency for
```
Comment (awni) [inline python/src/fast.cpp:98]: I'm thinking about making `base` an `Optional[float]` and making it an error to set both `base` and `freqs`. Wdyt?

File: mlx/backend/metal/rope.cpp (hunk 1 above)
Comment (awni) [reply mlx/backend/metal/rope.cpp:2]: Haven't done that in a while 😅

File: python/src/fast.cpp (hunk 2 above)
Comment (angeloskath) [reply python/src/fast.cpp:98]: Yeah I like it! The less ambiguity the better!

**Pull request merged**